### PR TITLE
MAINT: vendor Tempita in `scipy/_build_utils`

### PR DESCRIPTION
**Pull request by rgommers**
- Addresses a part of gh-20535.
- Purpose: make build more robust, stop relying on `import Cython` but only on invoking the `cython` executable
- Context for purpose: to make cross compilation better, we want to stop invoking the Python interpreter we're building for. Hence, we should avoid having to use/import things at build time outside of Python's stdlib. We're almost there, only a few places left where we invoke `custom_target(..., command: [py3, ...])` in `meson.build` files. Running `Tempita` is one of them.

This is an unmodified copy from `Cython/Tempita/` as of commit `023d4af35` in Cython's master branch (21 April 2024). Tempita hasn't been maintained independently on PyPI for 10+ years; we've used the Cython version which is semi-public (undocumented) for a while, with the note that we should vendor it if that ever fails. It now failed once (see scipy#20535), and conceptually that makes sense - this is the only place where we don't invoke the `cython` executable but actually do `import Cython`. That can fail in a distro setup where Cython is installed for a different Python interpreter than the active one.

**License**

Opening as a draft PR to clarify the licensing situation of this code first:
- The original `Tempita` was MIT-licensed, see https://pypi.org/project/Tempita/ (a single declaration, no full copy of the MIT license text in the sdist)
- That code was vendored in Cython in 2011, without making a note of the license: https://github.com/cython/cython/commits/master/Cython/Tempita
- Cython itself is Apache 2.0 licensed (not compatible with SciPy's BSD-3 license). _However_, there have been no substantial changes to the `Tempita` code since 2011 it looks like, but rather commits are minor maintenance/style change (running `2to3`, `pyupgrade`, etc.). So I think we should be able still consider this code as MIT-licensed - but that requires confirmation from @scoder I think, as Cython lead dev and author of most of the commits that touched this code. 
  - @scoder are you fine with including that `Cython/Tempita/` code under the original MIT license, or do you now consider it mixed MIT/Apache licensed?
  - Note that NumPy also vendored a copy until not long ago, and that does have an MIT license file: https://github.com/numpy/numpy/tree/maintenance/1.26.x/tools/npy_tempita. Using that would be the alternative. It is less well maintained than the Cython code, so we'd need to polish it a bit.

TODO:
- [x] Clarify license situation
- [x] Update `LICENSES_bundled.txt`

### CODE REVIEW

**Comment by scoder** [comment]
IANAL, but I looked through the changes and IMHO none of them would hand any non-trivial chunk of code over to other authors than in the original copy of Tempita that we included and possibly me. Almost all change were trivial adaptations or removals, many of them machine generated.

I hereby grant you the right to use my changes under the original code's MIT license, or the 3-clause BSD license that SciPy currently uses.

**Comment by rgommers** [comment]
Thanks @scoder, much appreciated!

**Comment by rgommers** [comment]
I knew the `lint` CI job was going to fail; `python dev.py lint` calls the `tools/lint.py --diff-against` which seems to not honor the `exclude` files in `lint.toml`. I'm not planning to reformat this code, so plan to just ignore the failure (it goes away by itself post-merge, at least until someone touches the file again). @lucascolley is this a known issue with the linter perhaps?

**Comment by lucascolley** [comment]
>  is this a known issue with the linter perhaps?

Right, I agree with your diagnosis. I can't think of an obvious refactor. It seems we would have to extract the `exclude` field from `lint.toml` in `lint.py` to perform set difference on the files. I'm happy to just let this fail.

File: tools/lint.toml [hunk 1]
```diff
@@ -1,4 +1,7 @@
-exclude = ["scipy/datasets/_registry.py"]
+exclude = [
+    "scipy/_build_utils/tempita/*",
+    "scipy/datasets/_registry.py",
+]
```
Comment (lucascolley) [inline tools/lint.toml:4]:
```suggestion
exclude = [
    "scipy/_build_utils/tempita/*",
    "scipy/datasets/_registry.py",
]

force-exclude=true
```

**Review by lucascolley — CHANGES_REQUESTED** [review]
never mind, here's the one-line (okay, two-line) fix!

**Review by lucascolley — APPROVED** [review]
LGTM!

**Comment by rgommers** [comment]
Thanks for the review @lucascolley. Going to hit the green button here, so I can work on the next step.

**Pull request merged**